### PR TITLE
Use friendly names

### LINKED ISSUE #966: [Feature] Simplify the resource naming of Liqo resources
So far, Liqo resources were named after the remote cluster id. To simplify the interaction with those resources, we would like to have pretty names for resources, which are easier to understand for users. This problem is particularly true when Liqo manages a lot of clusters, and it is hard to understand which cluster is involved.

Strategy: Liqo should use a unique local name to access resources locally; Liqo will use a hash of the UUID (max six characters) for remote disambiguation.

Tasks:

Rename LOCAL resources according to the new strategy starting from:

- [x] ForeignClusters
- [x] Nodes
- [x] NetworkConfig (from GenerateName)
- [x] TunnelEndpoint (from GenerateName)
- [x] Tenant
- [x] Tenant Namespaces
- [x] ResourceOffers (named after the cluster that created it)
- [x] ResourceRequests (named after the cluster that created it)

Rename REMOTE resources according to the new strategy starting from:
- [x] Application Namespaces

### PR DESCRIPTION
**Pull request by CapacitorSet**
# Description

Fixes #966.

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [ ] Unit tests
- [ ] E2E tests

### CODE REVIEW

File: apis/discovery/v1alpha1/foreigncluster_types.go [hunk 1]
```diff
@@ -15,6 +15,11 @@
 package v1alpha1
 
 import (
+
+	// nolint:gosec // we don't use md5 in the context of cryptography
+	"crypto/md5"
```
Comment (palexster) [inline apis/discovery/v1alpha1/foreigncluster_types.go:20]: Can we use sha256 instead of md5?

**Comment by CapacitorSet** [comment]
We can, but is there any benefit to it? In fact, we could directly use the
first few bytes of the cluster ID if it were guaranteed to be a UUID.

Il ven 26 nov 2021, 20:51 Alex Palesandro ***@***.***> ha
scritto:

File: apis/discovery/v1alpha1/foreigncluster_types.go (hunk 1 above)
Comment (CapacitorSet) [reply apis/discovery/v1alpha1/foreigncluster_types.go:20]: e can, but is there any benefit to it? In fact, we could directly use the
first few bytes of the cluster ID if it were guaranteed to be a UUID.

Comment (palexster) [reply apis/discovery/v1alpha1/foreigncluster_types.go:20]: Got it. However, I would avoid using md5 and all those nolint instructions.

File: install.sh [hunk 2]
```diff
@@ -416,10 +416,10 @@ function install_liqo() {
 	${KUBECTL} create namespace "${LIQO_NAMESPACE}" 1>/dev/null 2>&1 || true
 	local LIQO_CHART="${TMPDIR}/${LIQO_CHARTS_PATH}"
 	info "[INSTALL]" "Packaging the chart..."
-	${HELM} package --version="${LIQO_CHART_VERSION}" --app-version="${LIQO_CHART_VERSION}" "${LIQO_CHART}" --dependency-update >/dev/null ||
+	${HELM} --debug package --version="${LIQO_CHART_VERSION}" --app-version="${LIQO_CHART_VERSION}" "${LIQO_CHART}" --dependency-update >/dev/null ||
```
Comment (palexster) [inline install.sh:419]: This script will be deprecated. Why do you need to add --debug?

File: apis/discovery/v1alpha1/foreigncluster_types.go [hunk 3]
```diff
@@ -104,6 +109,18 @@ type ClusterIdentity struct {
 	ClusterName string `json:"clusterName"`
 }
 
+// UniqueName returns a user-friendly name that is unique *in the context of the current cluster*.
+// It depends on ClusterName, so the same cluster may have different UniqueNames in different clusters.
+//
+// Use it when reflecting resources on a remote cluster (see issue #966).
+func (i *ClusterIdentity) UniqueName() string {
```
Comment (palexster) [inline apis/discovery/v1alpha1/foreigncluster_types.go:116]: I think it should be defined elsewhere not in the API pkg.

File: internal/liqonet/network-manager/tunnelendpointcreator/tunnelendpointcreator.go [hunk 4]
```diff
@@ -382,10 +384,10 @@ func (tec *TunnelEndpointCreator) createTunnelEndpoint(ctx context.Context, para
 	// here we create it
 	tep := &netv1alpha1.TunnelEndpoint{
 		ObjectMeta: metav1.ObjectMeta{
-			GenerateName: tunEndpointNamePrefix,
+			GenerateName: fmt.Sprintf("tun-endpoint-%s-", param.remoteCluster.ClusterName),
```
Comment (palexster) [inline internal/liqonet/network-manager/tunnelendpointcreator/tunnelendpointcreator.go:387]: Since the resources are of type tun-endpoint maybe it is a bit repetitive to call them tun-endpoint?

File: internal/liqonet/network-manager/netcfgcreator/networkconfig.go [hunk 5]
```diff
@@ -127,7 +125,7 @@ func (ncc *NetworkConfigCreator) EnforceNetworkConfigPresence(ctx context.Contex
 func (ncc *NetworkConfigCreator) createNetworkConfig(ctx context.Context, fc *discoveryv1alpha1.ForeignCluster) error {
 	netcfg := netv1alpha1.NetworkConfig{
 		ObjectMeta: metav1.ObjectMeta{
-			GenerateName: networkConfigNamePrefix,
+			GenerateName: fmt.Sprintf("net-config-%s-", fc.Spec.ClusterIdentity.ClusterName),
```
Comment (palexster) [inline internal/liqonet/network-manager/netcfgcreator/networkconfig.go:128]: I would leave just the cluster name since the type is already known.

**Review by palexster — APPROVED** [review]

File: install.sh (hunk 2 above)
Comment (CapacitorSet) [reply install.sh:419]: It was for me to help with debugging (and as a possible improvement to be cherry-picked in a PR), I'll drop the commit.

**Comment by CapacitorSet** [comment]
Bug report: running `liqoctl install` twice with `--generate-name` reinstalls Liqo with a different name (on upstream, it does nothing).

File: pkg/virtualKubelet/reflection/workload/workload_suite_test.go [hunk 6]
```diff
@@ -40,14 +41,20 @@ const (
...
 
 var (
+	LocalCluster = discoveryv1alpha1.ClusterIdentity{
+		ClusterID:   "local-cluster-id",
+		ClusterName: "local-cluster",
+	}
+	RemoteCluster = discoveryv1alpha1.ClusterIdentity{
+		ClusterID:   "remote-cluster-id",
+		ClusterName: "remote-cluster",
+	}
```
Comment (giorio94) [inline pkg/virtualKubelet/reflection/workload/workload_suite_test.go:56]: Should be initialized in a BeforeEach block.

File: pkg/virtualKubelet/reflection/storage/storage_suite_test.go [hunk 7]
```diff
@@ -61,6 +59,15 @@ const (
...
 
 var (
+	LocalCluster = discoveryv1alpha1.ClusterIdentity{
+		ClusterID:   "local-cluster-id",
+		ClusterName: "local-cluster",
+	}
+	RemoteCluster = discoveryv1alpha1.ClusterIdentity{
+		ClusterID:   "remote-cluster-id",
+		ClusterName: "remote-cluster",
+	}
```
Comment (giorio94) [inline pkg/virtualKubelet/reflection/storage/storage_suite_test.go:69]: Same here

File: pkg/virtualKubelet/reflection/exposition/exposition_suite_test.go [hunk 8]
```diff
@@ -36,14 +37,20 @@ const (
...
 
 var (
+	LocalCluster = discoveryv1alpha1.ClusterIdentity{
+		ClusterID:   "local-cluster-id",
+		ClusterName: "local-cluster",
+	}
+	RemoteCluster = discoveryv1alpha1.ClusterIdentity{
+		ClusterID:   "remote-cluster-id",
+		ClusterName: "remote-cluster",
+	}
```
Comment (giorio94) [inline pkg/virtualKubelet/reflection/exposition/exposition_suite_test.go:52]: Same here

File: pkg/virtualKubelet/forge/forge_suite_test.go [hunk 9]
```diff
@@ -19,13 +19,24 @@ import (
...
+
+var (
+	LocalCluster = discoveryv1alpha1.ClusterIdentity{
+		ClusterID:   "local-cluster-id",
+		ClusterName: "local-cluster-name",
+	}
+	RemoteCluster = discoveryv1alpha1.ClusterIdentity{
+		ClusterID:   "remote-cluster-id",
+		ClusterName: "remote-cluster-name",
+	}
```
Comment (giorio94) [inline pkg/virtualKubelet/forge/forge_suite_test.go:39]: Same here

File: pkg/virtualKubelet/forge/endpointslices.go [hunk 10]
```diff
@@ -79,7 +78,10 @@ func RemoteEndpointSliceEndpoints(locals []discoveryv1beta1.Endpoint,
 
 		remote := discoveryv1beta1apply.Endpoint().
 			WithAddresses(translator(local.Addresses)...).WithConditions(conditions).
-			WithTopology(local.Topology).WithTopology(map[string]string{corev1.LabelHostname: nodename}).
+			WithTopology(local.Topology).WithTopology(map[string]string{
+			liqoconst.TypeLabel:       liqoconst.TypeNode,
+			liqoconst.RemoteClusterID: LocalCluster.ClusterID,
+		}).
```
Comment (giorio94) [inline pkg/virtualKubelet/forge/endpointslices.go:84]: Why is this modified?

File: pkg/tenantNamespace/namespaceManager.go [hunk 11]
```diff
@@ -120,3 +120,7 @@ func (nm *tenantNamespaceManager) GetNamespace(cluster discoveryv1alpha1.Cluster
 	}
 	return namespaces[0].DeepCopy(), nil
 }
+
+func getNameForNamespace(cluster discoveryv1alpha1.ClusterIdentity) string {
+	return fmt.Sprintf("liqo-tenant-%s", cluster.ClusterName)
```
Comment (giorio94) [inline pkg/tenantNamespace/namespaceManager.go:125]: Should this be constructed from the unique name?

File: internal/liqonet/network-manager/netcfgcreator/networkconfig_test.go [hunk 12]
```diff
@@ -242,7 +242,7 @@ var _ = Describe("Network config functions", func() {
 				It("the network config should be present and have the correct object meta", func() {
 					netcfg, err := GetLocalNetworkConfig(ctx, fcw.Client, clusterID, namespace)
 					Expect(err).ToNot(HaveOccurred())
-					Expect(netcfg.Name).To(HavePrefix("net-config-"))
+					Expect(netcfg.Name).To(Equal(clusterID)) // Equals cluster name
```
Comment (giorio94) [inline internal/liqonet/network-manager/netcfgcreator/networkconfig_test.go:245]: I'm missing the point of the comment.

File: pkg/liqo-controller-manager/foreign-cluster-operator/resourceRequest.go [hunk 13]
```diff
@@ -98,3 +98,7 @@ func resourceRequestLabels(remoteClusterID string) map[string]string {
 		consts.ReplicationDestinationLabel: remoteClusterID,
 	}
 }
+
+func getResourceRequestNameFor(cluster discoveryv1alpha1.ClusterIdentity) string {
+	return fmt.Sprintf("resourcerequest-%s", cluster.ClusterName)
+}
```
Comment (giorio94) [inline pkg/liqo-controller-manager/foreign-cluster-operator/resourceRequest.go:104]: I feel it is probably useless to prepend `resourcerequest`, given that is implied by the resource type. Should the name be based on the unique name version?

File: pkg/liqo-controller-manager/resource-request-controller/resourceRequest_controller_methods.go [hunk 14]
```diff
@@ -129,3 +128,13 @@ func (r *ResourceRequestReconciler) checkOfferState(ctx context.Context,
...
+
+// getOfferName returns the name of the ResourceOffer coming from the given cluster.
+func getOfferName(cluster discoveryv1alpha1.ClusterIdentity) string {
+	return fmt.Sprintf("resourceoffer-%s", cluster.ClusterName)
+}
+
+// getTenantName returns the name of the Tenant for the given cluster.
+func getTenantName(cluster discoveryv1alpha1.ClusterIdentity) string {
+	return fmt.Sprintf("tenant-%s", cluster.ClusterName)
+}
```
Comment (giorio94) [inline pkg/liqo-controller-manager/resource-request-controller/resourceRequest_controller_methods.go:140]: Same here

File: pkg/liqoctl/generate/generate_test.go [hunk 15]
```diff
@@ -72,6 +73,7 @@ var _ = Describe("Test the generate command works as expected", func() {
 			},
 			Data: map[string]string{
 				consts.ClusterIDConfigMapKey: localClusterID,
+				"CLUSTER_NAME":               localClusterName,
```
Comment (giorio94) [inline pkg/liqoctl/generate/generate_test.go:76]: Should this be a constant?

File: pkg/liqo-controller-manager/resource-request-controller/utils.go [hunk 16]
```diff
@@ -72,7 +72,8 @@ func (r *ResourceRequestReconciler) createForeignCluster(ctx context.Context,
 	resourceRequest *discoveryv1alpha1.ResourceRequest) (*discoveryv1alpha1.ForeignCluster, error) {
 	foreignCluster := &discoveryv1alpha1.ForeignCluster{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: resourceRequest.Spec.ClusterIdentity.ClusterID,
+			// The resource request was created locally by the user with `liqoctl add`, so we can use ClusterName
```
Comment (giorio94) [inline pkg/liqo-controller-manager/resource-request-controller/utils.go:75]: AFAIK The resource request is not created by `liqoctl`

File: pkg/liqo-controller-manager/namespaceMap-controller/ensure_remote_namespaces.go [hunk 17]
```diff
@@ -228,28 +228,40 @@ func (r *NamespaceMapReconciler) namespaceMapDeletionProcess(ctx context.Context
 
 // checkRemoteNamespaceRoleBindings checks that the right roleBindings are inside the remote namespace to understand if the
 // virtual kubelet will have the right privileges on that namespace.
-func checkRemoteNamespaceRoleBindings(ctx context.Context, cl kubernetes.Interface,
```
Comment (giorio94) [inline pkg/liqo-controller-manager/namespaceMap-controller/ensure_remote_namespaces.go:231]: I am personally missing why this change is required.

**Review by giorio94 — COMMENTED** [review]
Some comments inline.

I am also wondering whether it is useful to use the cluster identity in the virtual kubelet, given it adds complexity and it seems to me the cluster name is never used.

File: pkg/virtualKubelet/forge/endpointslices.go (hunk 10 above)
Comment (CapacitorSet) [reply pkg/virtualKubelet/forge/endpointslices.go:84]: Because the node name now depends on the cluster name, which in turn may vary across clusters. This change rewrites the query to be independent of names.

File: pkg/liqo-controller-manager/foreign-cluster-operator/resourceRequest.go (hunk 13 above)
Comment (CapacitorSet) [reply pkg/liqo-controller-manager/foreign-cluster-operator/resourceRequest.go:104]: Resource requests are created in the tenant namespace, so I believe in principle they might as well be named `request`. I'll need to check that.

File: pkg/virtualKubelet/forge/endpointslices.go (hunk 10 above)
Comment (giorio94) [reply pkg/virtualKubelet/forge/endpointslices.go:84]: Ok, but that is not the purpose of that field (https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#topology).

BTW, please do not mark as resolved conversations which are not effectively resolved.

File: pkg/liqo-controller-manager/resource-request-controller/resourceRequest_controller_methods.go (hunk 14 above)
Comment (CapacitorSet) [reply pkg/liqo-controller-manager/resource-request-controller/resourceRequest_controller_methods.go:140]: The UniqueName is not needed for the tenant, but for it to be safe it must be derived from the ClusterIdentity for the local ForeignCluster, not the ClusterIdentity provided in the ResourceRequest. I fixed that, thanks.

I also removed the prefix.

File: pkg/tenantNamespace/namespaceManager.go (hunk 11 above)
Comment (CapacitorSet) [reply pkg/tenantNamespace/namespaceManager.go:125]: Yes, good catch - the tenant ns may be created as part of the auth process, when the cluster name is not guaranteed to be unique.

File: pkg/liqo-controller-manager/namespaceMap-controller/ensure_remote_namespaces.go (hunk 17 above)
Comment (CapacitorSet) [reply pkg/liqo-controller-manager/namespaceMap-controller/ensure_remote_namespaces.go:231]: As for the first few rows: because the role binding labels refers to the tenant, which depends on the cluster name set by (I believe) the remote cluster, which is unknown (it is not necessarily the same as the cluster name of the local installation).

For the second half, I figured listing the role bindings explicitly was much clearer than expecting 3 bindings.

File: apis/discovery/v1alpha1/foreigncluster_types.go (hunk 3 above)
Comment (CapacitorSet) [reply apis/discovery/v1alpha1/foreigncluster_types.go:114]: This issue is still un-addressed. Do you think there is a better package for these?

**Comment by CapacitorSet** [comment]
Should now be ok expect for the additional unit test that I can't get right. I did eventually remove cluster identities from the vk.

File: pkg/liqo-controller-manager/namespaceMap-controller/ensure_remote_namespaces.go (hunk 17 above)
Comment (CapacitorSet) [reply pkg/liqo-controller-manager/namespaceMap-controller/ensure_remote_namespaces.go:231]: Can this conversation be marked as resolved?

File: pkg/virtualKubelet/liqoNodeProvider/setup.go [hunk 18]
```diff
@@ -80,7 +80,7 @@ func NewLiqoNodeProvider(cfg *InitConfig) *LiqoNodeProvider {
 
 func node(cfg *InitConfig) *corev1.Node {
 	lbls := map[string]string{
-		corev1.LabelHostname: cfg.NodeName,
+		corev1.LabelHostname: cfg.RemoteClusterID,
```
Comment (giorio94) [inline pkg/virtualKubelet/liqoNodeProvider/setup.go:83]: Why is this change required?

File: pkg/liqo-controller-manager/storageprovisioner/remoteprovisioner.go [hunk 19]
```diff
@@ -77,9 +78,9 @@ func ProvisionRemotePVC(ctx context.Context,
 						{
 							MatchExpressions: []corev1.NodeSelectorRequirement{
 								{
-									Key:      corev1.LabelHostname,
+									Key:      consts.RemoteClusterID,
 									Operator: corev1.NodeSelectorOpIn,
-									Values:   []string{options.SelectedNode.Name},
+									Values:   []string{options.SelectedNode.Labels[consts.RemoteClusterID]},
```
Comment (giorio94) [inline pkg/liqo-controller-manager/storageprovisioner/remoteprovisioner.go:83]: Why is this change required?

File: pkg/virtualKubelet/liqoNodeProvider/setup.go (hunk 18 above)
Comment (CapacitorSet) [reply pkg/virtualKubelet/liqoNodeProvider/setup.go]: On a closer look it's not actually required - I reverted it.

File: apis/discovery/v1alpha1/foreigncluster_types.go [hunk 20]
```diff
@@ -107,6 +110,23 @@ type ClusterIdentity struct {
 	ClusterName string `json:"clusterName"`
 }
 
+// UniqueName returns a user-friendly name that is unique *in the context of the current cluster*.
```
Comment (aleoli) [inline apis/discovery/v1alpha1/foreigncluster_types.go:113]: Methods should be moved to the foreigncluster utils package, in the API we should only have types definitions

The method can have a signature like
```go
func UniqueName(*ClusterIdentity) string
```

File: internal/auth-service/identity.go [hunk 21]
```diff
@@ -95,12 +94,9 @@ func (authService *Controller) handleIdentity(
 	}
 	tracer.Step("Credentials checked")
 
-	klog.V(4).Infof("Creating Tenant Namespace for cluster %v", identityRequest.GetClusterID())
-	clusterIdentity := discoveryv1alpha1.ClusterIdentity{
-		ClusterID:   identityRequest.ClusterID,
-		ClusterName: identityRequest.ClusterID,
-	}
-	namespace, err := authService.namespaceManager.CreateNamespace(clusterIdentity)
+	cluster := identityRequest.ClusterIdentity
```
Comment (aleoli) [inline internal/auth-service/identity.go:97]:
```suggestion
	remoteClusterIdentity := identityRequest.ClusterIdentity
```

File: apis/net/v1alpha1/networkconfig_types.go [hunk 22]
```diff
@@ -26,8 +28,8 @@ type NetworkConfigSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 
-	// The ID of the remote cluster that will receive this CRD.
-	ClusterID string `json:"clusterID"`
+	// The remote cluster that will receive this CRD.
+	Cluster discoveryv1alpha1.ClusterIdentity `json:"cluster"`
```
Comment (aleoli) [inline apis/net/v1alpha1/networkconfig_types.go:32]:
```suggestion
	RemoteCluster discoveryv1alpha1.ClusterIdentity `json:"cluster"`
```

File: internal/liqonet/network-manager/netcfgcreator/netcfgcreator_test.go [hunk 23]
```diff
@@ -143,7 +147,7 @@ var _ = Describe("NetworkConfigCreator Controller", func() {
 				}
 
 				AssertNetworkConfigSpec := func(netcfg *netv1alpha1.NetworkConfig) {
-					Expect(netcfg.Spec.ClusterID).To(BeIdenticalTo(clusterID))
+					Expect(netcfg.Spec.Cluster.ClusterID).To(BeIdenticalTo(clusterID))
```
Comment (aleoli) [inline internal/liqonet/network-manager/netcfgcreator/netcfgcreator_test.go:150]:
```suggestion
					Expect(netcfg.Spec.Cluster.ClusterID).To(BeIdenticalTo(clusterID))
					Expect(netcfg.Spec.Cluster.ClusterName).To(BeIdenticalTo(clusterName))
```

File: pkg/auth/identityrequest.go [hunk 24]
```diff
@@ -14,24 +14,28 @@
...
-	ClusterID string `json:"clusterID"`
-	Token     string `json:"token"`
+	ClusterIdentity discoveryv1alpha1.ClusterIdentity `json:"cluster"`
+	Token           string                            `json:"token"`
 }
 
 // CertificateIdentityRequest is the request for a new certificate validation.
 type CertificateIdentityRequest struct {
-	ClusterID string `json:"clusterID"`
+	ClusterIdentity discoveryv1alpha1.ClusterIdentity `json:"cluster"`
```
Comment (aleoli) [inline pkg/auth/identityrequest.go:38]: Is this change required? The certificate should contain the unique clusterID only

File: pkg/liqo-controller-manager/resource-request-controller/utils.go [hunk 25]
```diff
@@ -68,18 +68,31 @@ func (r *ResourceRequestReconciler) ensureControllerReference(foreignCluster *di
 	return true, controllerutil.SetControllerReference(foreignCluster, resourceRequest, r.Scheme)
 }
 
+// createForeignCluster creates the foreign cluster for the given RR. It tries using the cluster name first, and retries
+// with the UniqueName if there is already a foreign cluster with the same name.
 func (r *ResourceRequestReconciler) createForeignCluster(ctx context.Context,
-	resourceRequest *discoveryv1alpha1.ResourceRequest) (*discoveryv1alpha1.ForeignCluster, error) {
+	resourceRequest *discoveryv1alpha1.ResourceRequest, isRetry bool) (*discoveryv1alpha1.ForeignCluster, error) {
```
Comment (aleoli) [inline pkg/liqo-controller-manager/resource-request-controller/utils.go:74]: You should avoid using hard-to-read bool values

You can wrap this function like that
```go
func createForeignCluster(context.Context, *discoveryv1alpha1.ResourceRequest) (*discoveryv1alpha1.ForeignCluster, error) {
  return createForeignClusterWithRetry(ctx, rr, false)
}

func createForeignClusterWithRetry(context.Context, *discoveryv1alpha1.ResourceRequest, isRetry bool) (*discoveryv1alpha1.ForeignCluster, error)
```

File: pkg/virtualKubelet/forge/endpointslices.go [hunk 26]
```diff
@@ -66,7 +64,7 @@ func RemoteEndpointSlice(local *discoveryv1beta1.EndpointSlice, targetNamespace
 func RemoteEndpointSliceEndpoints(locals []discoveryv1beta1.Endpoint,
 	translator EndpointTranslator) []*discoveryv1beta1apply.EndpointApplyConfiguration {
 	var remotes []*discoveryv1beta1apply.EndpointApplyConfiguration
-	nodename := virtualKubelet.VirtualNodeName(LocalClusterID)
+	hostname := LocalClusterID
```
Comment (aleoli) [inline pkg/virtualKubelet/forge/endpointslices.go:67]: Is this correct? Shouldn't it be the same as the virtual node name?

**Review by aleoli — APPROVED** [review]
